### PR TITLE
Explore: Change datasourceName usage to datasourceUID or datasourceRef where appropriate

### LINKED ISSUE #49496: Refactor usage of datasourceId and/or datasourceName in favor of datasource UID
Find all usages of datasourceId / datasourceName and investigate if we can switch over to using a datasource ref or datasource UID.

- Change Rich History files to pass references to datasources by UID. They will still use datasource name for display purposes. Rich history entries are already saved with datasource UIDs.
- Change URL state to use UID but support lookup by name to be backwards compatible. We do not plan on deprecating non UID URLs because the root datasource lookup functions support both name and UID for lookup. We will look into changing this if/when the datasource service does.
- `TraceView` uses datasource name to generate the title for a data link if one is not provided. We will use the UID for the URL itself
- Change `ExploreToolbar` to pass uid to the datasource picker
- Fix relevant tests

### PR DESCRIPTION
**Pull request by gelicia**
In explore, remove usage of datasourceName when we can be referring to a datasource by UID. 

The URLs to explore will for the foreseeable future support both datasource name and datasource UID. Since the root lookup is a get call that takes in strings and refs and will try to make a match on whatever ([src](https://github.com/grafana/grafana/blob/main/public/app/features/plugins/datasource_srv.ts#L115)) we will keep that logic for the initialization logic Explore uses. 

Fixes #49496

### CODE REVIEW

File: public/app/features/explore/state/utils.ts [hunk 1]
```diff
@@ -86,11 +87,12 @@ export const createEmptyQueryResponse = (): ExplorePanelData => ({
 
 export async function loadAndInitDatasource(
   orgId: number,
-  datasourceUid?: string
```
Comment (gelicia) [inline public/app/features/explore/state/utils.ts:89]: This still gets called from places where the URL might send in a name as a string, so we allow both types here. Basically we should let this be equivalent to whatever `datasourceSrv.get` accepts.

File: public/app/features/explore/state/utils.ts [hunk 2]
```diff
@@ -86,11 +87,12 @@ export const createEmptyQueryResponse = (): ExplorePanelData => ({
...
+  datasource: DataSourceRef | string
```
Comment (gelicia) [inline public/app/features/explore/state/utils.ts:90]: This gets called from a wider initialize Explore call where the URL might send in a name as a string, so we allow both types here too. 

Basically we should let this be equivalent to whatever datasourceSrv.get accepts.

File: packages/grafana-data/src/field/fieldOverrides.test.ts [hunk 3]
```diff
@@ -645,7 +645,7 @@ describe('getLinksSupplier', () => {
     expect(links[0]).toEqual(
       expect.objectContaining({
         title: 'testDS',
-        href: `/explore?left=${encodeURIComponent('{"datasource":"testDS","queries":["12345"],"panelsState":{}}')}`,
+        href: `/explore?left=${encodeURIComponent('{"datasource":"1234","queries":["12345"],"panelsState":{}}')}`,
```
Comment (leventebalogh) [inline packages/grafana-data/src/field/fieldOverrides.test.ts:648]: **Nit:** maybe we could extract the `datasourceUid` to a separate variable to remove some duplication
```javascript
const datasourceUid = "1234";
const f0 = new MutableDataFrame({
// ...
```

File: public/app/features/explore/state/explorePane.ts [hunk 4]
```diff
@@ -138,7 +139,7 @@ export function changeGraphStyle(exploreId: ExploreId, graphStyle: ExploreGraphS
  */
 export function initializeExplore(
   exploreId: ExploreId,
-  datasourceNameOrUid: string,
+  datasource: DataSourceRef | string,
```
[the remark this answers is not in the document]
Comment (leventebalogh) [reply public/app/features/explore/state/explorePane.ts:142]: Although this makes sense, maybe it worth to add this information somehow as a short comment, so it's easier to understand later why this type is a bit ambiguous. (I am saying this even though I am generally against comments 😛)

File: public/app/features/explore/state/main.ts [hunk 5]
```diff
@@ -108,9 +108,9 @@ export const splitOpen: SplitOpen = (options): ThunkResult<void> => {
     let rightUrlState: ExploreUrlState = leftUrlState;
 
     if (options) {
-      const datasourceName = getDataSourceSrv().getInstanceSettings(options.datasourceUid)?.name || '';
+      const datasourceUid = getDataSourceSrv().getInstanceSettings(options.datasourceUid)?.uid || '';
```
Comment (leventebalogh) [inline public/app/features/explore/state/main.ts:111]: **Suggestion:** I think we can just use `options.datasourceUid` straight away?

File: packages/grafana-data/src/types/dataLink.ts [hunk 6]
```diff
@@ -44,7 +44,7 @@ export interface DataLink<T extends DataQuery = any> {
 export interface InternalDataLink<T extends DataQuery = any> {
   query: T;
   datasourceUid: string;
-  datasourceName: string;
+  datasourceName: string; // used as a title if one is not defined
```
Comment (leventebalogh) [inline packages/grafana-data/src/types/dataLink.ts:47]: **Suggestion:** I like this addition, although I would probably rephrase it a little bit to make it clearer  
```suggestion
  datasourceName: string; // used as a title if `DataLink.title` is empty
```

File: public/app/features/explore/state/main.ts (hunk 5 above)
Comment (gelicia) [reply public/app/features/explore/state/main.ts:111]: 🤦‍♀️ thank you!

**Review by kaydelaney — APPROVED** [review]

File: packages/grafana-data/src/types/dataLink.ts (hunk 6 above)
Comment (gelicia) [reply packages/grafana-data/src/types/dataLink.ts:47]: @leventebalogh actually looking at this, I think we can remove it altogether. 

logic that uses datasourceName: https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/utils/dataLinks.ts#L48

File: public/app/features/explore/state/explorePane.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/state/explorePane.ts:146]: I added a comment above because I couldnt figure out how to be more succinct with it.

**Review by leventebalogh — APPROVED** [review]

**Pull request merged**